src/lib/util/FileHandler.ts

```diff
@@ -109,15 +111,45 @@ export class FileHandler {
 		});
 	}
 
-	linkFile(src: string, target: string): Promise<void> {
+	duplicateFile(src: string, target: string): Promise<void> {
+		// NOTE: This method is likely to be called many times concurrently, and it's probably possible to optimize it a
+		// bit so we don't try duplication methods we know will fail. However, that introduces risk that we'd rather avoid
+		// at the moment. So we'll go with this semi-naive implementation, aware that it performs SLIGHTLY worse than
+		// an optimal one, and prepared to address it if there's somehow a problem.
+		// These are the file duplication functions available to us, in order of preference.
+		const dupFns: DuplicationFn[] = [fs.symlink, fs.link, fs.copyFile];
+		const errs: Error[] = [];
+
 		return new Promise<void>((res, rej) => {
-			fs.link(src, target, (err) => {
-				if (err) {
-					rej(err);
+			const dfIterator: IterableIterator<DuplicationFn> = dupFns.values();
+
+			function iterativelyAttemptDuplication(): void {
+				const {value, done} = dfIterator.next();
+				if (done) {
+					// If `done` is true, it means we're out of duplication functions to try. We'll need to combine the
+					// error messages we got in order to make something informative and helpful.
+					const errMsgs: string[] = [];
+					for (let i = 0; i < errs.length; i++) {
```

Comment:
I was trying to determine if the indexes could get out of sync, I'm not sure if they can, but it seems that this array could be built inside of the `err` block instead of keeping two separate arrays and then converting them to strings.

Line 121 would change to `const errMsgs: string[] = [];` and 
Line 141 would change to `errMsgs.push(`${value.name}: ${err.message || err}`);`

Reply:
That works too. One sec.